### DISCUSSION TITLE
Distributed mesh with element based material properties

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#25440] [Q&A General]
Hi guys,

I'm trying to moved my simulation into distributed mesh to reduce RAM usage (as we approaching the upper limit)

so I got an error at the `PropertyReadFile`

```
*** ERROR ***
The following error occurred in the object "flag_read_MDLCHG_FLAG_0039", of type "PropertyReadFile".

Element ID 36759009 greater than than total number of element in mesh: 9357805. Elements should be numbered consecutively.



*** ERROR ***
The following error occurred in the object "flag_read_MDLCHG_FLAG_0039", of type "PropertyReadFile".

Element ID 27351999 greater than than total number of element in mesh: 9357805. Elements should be numbered consecutively.



*** ERROR ***
The following error occurred in the object "flag_read_MDLCHG_FLAG_0039", of type "PropertyReadFile".

Element ID 16219069 greater than than total number of element in mesh: 9357805. Elements should be numbered consecutively.

```

I run it with 4 cores, i guess Element ID are the multiple of 2 3 4 from an original ID?

I checked the source code:

it seems like the total Element ID is multiple by a number of cores?

```
Real
PropertyReadFile::getElementData(const Elem * elem, unsigned int prop_num) const
{
  unsigned int jelem = elem->id();
  if (jelem >= _mesh.nElem())
    mooseError("Element ID ",
               jelem,
               " greater than than total number of element in mesh: ",
               _mesh.nElem(),
               ". Elements should be numbered consecutively.");
  return _reader.getData(jelem)[prop_num];
}
```

I even has:
```
[Mesh] 
allow_renumbering = false
```

is this a bug? because` jelem = elem->id();` should not add extra element ID?

Kind regards,
Traiwit

**Comment by GiudGiud** [comment]
Hello

i think you should allow_renumbering to make sure to get consecutive element ids.

i m wondering if we should force a renumbering step even 

Are you using  adaptivity?

Guillaume

**Reply by GiudGiud** [reply]
How are you mapping the ids and the elements when preparing the CSV file? You should have one value for each element but that seems tough to do with an arbitrary element numbering

**Reply by Traiwit** [reply]
> i think you should allow_renumbering to make sure to get consecutive element ids.

Hi @GiudGiud, i dont think so, the code itself even has a warning in there:

```
void
PropertyReadFile::readData()
{
  if (_read_type == ReadTypeEnum::ELEMENT && _mesh.getMesh().allow_renumbering())
    mooseWarning("CSV data is sorted by element, but mesh element renumbering is on, be careful!");

  _reader.setFormatFlag(MooseUtils::DelimitedFileReader::FormatFlag::ROWS);
  _reader.read();
```

in the normal case it works perfectly fine, so each line of the .csv is the property value of each element_ID from 1 to n

for example

0
0
1e-12
2e-12
0

(so element ID 3 and 4 has value, the rest is 0)

**Reply by Traiwit** [reply]
note that, in the normal case, I have element ID goes from 1 to n, so that way I can match the .csv with it

**Reply by GiudGiud** [reply]
I added the warning about renumbering because I don’t see how someone can predict how it will be renumbered.

but if you feel you can work with the post-renumbering ids you can try that

**Reply by GiudGiud** [reply]
> I run it with 4 cores, i guess Element ID are the multiple of 2 3 4 from an original ID?

is that how the numbering works for a distrivuted mesh that is loaded from file  @roystgnr

**Reply by Traiwit** [reply]
>I added the warning about renumbering because I don’t see how someone can predict how it will be renumbered.

>but if you feel you can work with the post-renumbering ids you can try that

yep I turn it off as well, as mentioned above


but yeh now the element ID is kinda more than the total element number, which is kinda weird.

maybe I did it incorrectly? I just added `--distributed-mesh` at the end of the mpiexec line:

mpiexec -n 4 ~/moose_beh/moose_beh-opt -i XXXXXXXX_R04H_M02_Nocap_transient_part3.i --distributed-mesh --t --color off --n-threads=4 

here is my mesh block:

```

### ################################################################### #######
### MESH (NODE, ELEMENTS) AND
### SETS (NSETS, ELSETS, SRFSETS,) AS WELL AS MPCS
### ################################################################### #######
[Mesh] 
allow_renumbering = false
[mesh]
type = FileMeshGenerator
allow_renumbering = false
file= XXXXXXXXXXX_G01HTv1_FT_MOOSEMESH_updated.inp
[]
### (optional) parsing sets
[side_south]
   type = ParsedGenerateSideset
   input = mesh
   combinatorial_geometry = 'y<=-1607.6' 
   new_sideset_name = 'BC_WALL_SOUTH' 
 []
 [side_north]
    type = ParsedGenerateSideset
    input = side_south
    combinatorial_geometry = 'y>=7772.2' 
    new_sideset_name = 'BC_WALL_NORTH' 
  []
  [side_east]
     type = ParsedGenerateSideset
     input = side_north
     combinatorial_geometry = 'x>=8520.4' 
     new_sideset_name = 'BC_WALL_EAST' 
   []
   [side_west]
      type = ParsedGenerateSideset
      input = side_east
      combinatorial_geometry = 'x<=-859.47' 
      new_sideset_name = 'BC_WALL_WEST' 
    []
[]

```

**Reply by roystgnr** [reply]
> is that how the numbering works for a distrivuted mesh that is loaded from file

Not on purpose.  With renumbering disabled, the ids ought be the original ids (modulo whatever we need to get the file format to be happy with them, e.g. subtracting 1 to go from 1-based ExodusII counting to 0-base libMesh counting).  With renumbering enabled, the ids are going to end up sorted onto contiguous blocks on processors and aren't going to have much to do with the original ids.

But this is an Abaqus file?  I don't know how much love AbaqusIO has received, especially post-DistributedMesh.  I'm seeing us setting up an `_abaqus_to_libmesh_elem_mapping` as we add libMesh elements, for instance, rather than just simply pre-setting the libMesh element id to the Abaqus id like it should.  In *that* case, what's going to happen is that `DistributedMesh::add_elem` will try to pick an id in such a way that the synchronized addition of these elements can't possibly conflict with any possible parallel addition of other elements; it'll leave spaces in between the ids so that each processor can add its own elements as needed without a collision.  With N processors, the first added element ought to be getting id N+1, the next 2N+2, etc.

This ought to be a quick fix in libMesh abaqus_io.C, if @Traiwit can help me test it, but it'll probably be a month before the changes make it into MOOSE.

**Reply by roystgnr** [reply]
Actually - do we know the range of Abaqus-allowed element numbers?  I'm assuming it's 1-based, and we'll want to do the same conversion to 0-based that we do with Exodus?

**Reply by roystgnr** [reply]
Anyway, I'm just about done with work for the day, but I threw up https://github.com/libMesh/libmesh/pull/3640 if you want to try it out.

**Reply by Traiwit** [reply]
Hi @roystgnr @GiudGiud 

How do I even modify libmesh from my end? note that I installed my MOOSE via mamba, I dont have access to go inside libmesh 

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
Hi guys, i'm not super sure what to do:

so I followed this (below), but installed it inside moose/libmesh

https://github.com/libMesh/libmesh

then should I just follow:

https://mooseframework.inl.gov/help/faq/faq_build_libmesh.html

Regards,
Traiwit

**Reply by GiudGiud** [reply]
Hello

So I would just create another moose folder (just to keep your old one working)
Then 
- deactivate the moose environment
- go in moose/scripts
- run update_and_rebuild_petsc
- set LIBMESH_DIR to a libmesh directory set to the right submodule
- (git remote add roystgnr git@github.com:roystgnr/libmesh.git; git fetch roystgnr; git checkout --track roystgnr/abaqus_numbering)
- run update_and_rebuild_libmesh --skip-submodule-update

Guillaume

**Reply by roystgnr** [reply]
And for the "libmesh directory set to the right submodule" part: in `moose/libmesh` you'll want to `git remote add roystgnr git@github.com:roystgnr/libmesh.git`; `git fetch roystgnr`; `git checkout --track roystgnr/abaqus_numbering`
